File: sdk/typescript/tests/adp_criticality.test.ts
/**
 * ADP REPLACE-criticality wiring (spec §6 / §7).
 *
 * REPLACE operations on the ASD must be transmitted with FLAG_CRITICAL
 * because a lost REPLACE leaves the receiving node with a stale dictionary
 * entry — a semantic correctness violation. These tests lock in the
 * send-side helper (`deltaRequiredOverflowFlags`) and the receive-side
 * validator (`validateReceivedDelta` + `DeltaValidationError`).
 *
 * License: Apache 2.0
 */
import { describe, it, expect } from "vitest";
import {
  ADPDelta,
  deltaRequiredOverflowFlags,
  validateReceivedDelta,
  DeltaValidationError,
} from "../src/adp.js";
import { FLAG_CRITICAL } from "../src/types.js";

const ADDITIVE_DELTA: ADPDelta = {
  fromVersion: "15.0",
  toVersion: "15.1",
  operations: [
    { namespace: "H", mode: "+", opcode: "LACTATE", definition: "lactate" },
  ],
};

const DEPRECATE_DELTA: ADPDelta = {
  fromVersion: "15.0",
  toVersion: "15.1",
  operations: [{ namespace: "M", mode: "†", opcode: "OLDOP" }],
};

const REPLACE_DELTA: ADPDelta = {
  fromVersion: "15.0",
  toVersion: "15.1",
  operations: [
    { namespace: "H", mode: "←", opcode: "HR", definition: "heart_rate_v2" },
  ],
};

const MIXED_DELTA: ADPDelta = {
  fromVersion: "15.0",
  toVersion: "15.1",
  operations: [
    { namespace: "H", mode: "+", opcode: "LACTATE" },
    { namespace: "H", mode: "←", opcode: "HR", definition: "v2" },
  ],
};

describe("deltaRequiredOverflowFlags", () => {
  it("returns 0 for additive-only delta", () => {
    expect(deltaRequiredOverflowFlags(ADDITIVE_DELTA)).toBe(0);
  });

  it("returns 0 for deprecate-only delta", () => {
    expect(deltaRequiredOverflowFlags(DEPRECATE_DELTA)).toBe(0);
  });

  it("returns FLAG_CRITICAL for REPLACE delta", () => {
    expect(deltaRequiredOverflowFlags(REPLACE_DELTA)).toBe(FLAG_CRITICAL);
  });

  it("returns FLAG_CRITICAL when any operation is REPLACE", () => {
    expect(deltaRequiredOverflowFlags(MIXED_DELTA)).toBe(FLAG_CRITICAL);
  });
});

describe("validateReceivedDelta", () => {
  it("rejects REPLACE delta arriving without FLAG_CRITICAL", () => {
    const sal = "A:ASD:DELTA[15.0→15.1:H←[HR]]";
    expect(() => validateReceivedDelta(sal, 0)).toThrow(DeltaValidationError);
  });

  it("accepts REPLACE delta arriving with FLAG_CRITICAL set", () => {
    const sal = "A:ASD:DELTA[15.0→15.1:H←[HR]]";
    expect(() => validateReceivedDelta(sal, FLAG_CRITICAL)).not.toThrow();
  });

  it("accepts additive delta arriving without FLAG_CRITICAL", () => {
    const sal = "A:ASD:DELTA[15.0→15.1:H+[LACTATE]]";
    expect(() => validateReceivedDelta(sal, 0)).not.toThrow();
  });

  it("passes through non-DELTA SAL", () => {
    expect(() => validateReceivedDelta("H:HR@NODE1", 0)).not.toThrow();
    expect(() => validateReceivedDelta("A:ASD[15.1]", 0)).not.toThrow();
    expect(() => validateReceivedDelta("A:ASD?", 0)).not.toThrow();
  });
});

File: sdk/go/osmp/adp_criticality_test.go
package osmp

import "testing"

// TestRequiredOverflowFlagsAdditive verifies an additive-only delta does
// not require the criticality flag.
func TestRequiredOverflowFlagsAdditive(t *testing.T) {
	d := &ADPDelta{
		FromVersion: "15.0", ToVersion: "15.1",
		Operations: []ADPDeltaOp{{
			Namespace: "H", Mode: "+", Opcode: "LACTATE", Definition: "lactate",
		}},
	}
	if got := d.RequiredOverflowFlags(); got != 0 {
		t.Fatalf("additive delta: want 0, got 0x%02x", got)
	}
}

// TestRequiredOverflowFlagsDeprecate verifies deprecate deltas do not
// require the criticality flag.
func TestRequiredOverflowFlagsDeprecate(t *testing.T) {
	d := &ADPDelta{
		FromVersion: "15.0", ToVersion: "15.1",
		Operations: []ADPDeltaOp{{
			Namespace: "M", Mode: "†", Opcode: "OLDOP",
		}},
	}
	if got := d.RequiredOverflowFlags(); got != 0 {
		t.Fatalf("deprecate delta: want 0, got 0x%02x", got)
	}
}

// TestRequiredOverflowFlagsReplace verifies REPLACE deltas require the
// criticality flag (spec §6 / §7 mandatory retransmission).
func TestRequiredOverflowFlagsReplace(t *testing.T) {
	d := &ADPDelta{
		FromVersion: "15.0", ToVersion: "15.1",
		Operations: []ADPDeltaOp{{
			Namespace: "H", Mode: "←", Opcode: "HR", Definition: "heart_rate_v2",
		}},
	}
	if got := d.RequiredOverflowFlags(); got != FlagCritical {
		t.Fatalf("REPLACE delta: want FlagCritical (0x%02x), got 0x%02x",
			FlagCritical, got)
	}
}

// TestRequiredOverflowFlagsMixed verifies a mixed delta with at least one
// REPLACE op requires criticality even when other ops are additive.
func TestRequiredOverflowFlagsMixed(t *testing.T) {
	d := &ADPDelta{
		FromVersion: "15.0", ToVersion: "15.1",
		Operations: []ADPDeltaOp{
			{Namespace: "H", Mode: "+", Opcode: "LACTATE"},
			{Namespace: "H", Mode: "←", Opcode: "HR", Definition: "v2"},
		},
	}
	if got := d.RequiredOverflowFlags(); got != FlagCritical {
		t.Fatalf("mixed delta: want FlagCritical, got 0x%02x", got)
	}
}

// TestValidateReceivedDeltaReplaceWithoutCriticalRejects verifies a REPLACE
// delta arriving without FlagCritical is rejected.
func TestValidateReceivedDeltaReplaceWithoutCriticalRejects(t *testing.T) {
	sal := "A:ASD:DELTA[15.0→15.1:H←[HR]]"
	if err := ValidateReceivedDelta(sal, 0); err == nil {
		t.Fatal("expected error on REPLACE without FlagCritical")
	}
}

// TestValidateReceivedDeltaReplaceWithCriticalAccepts verifies a REPLACE
// delta with FlagCritical set is accepted.
func TestValidateReceivedDeltaReplaceWithCriticalAccepts(t *testing.T) {
	sal := "A:ASD:DELTA[15.0→15.1:H←[HR]]"
	if err := ValidateReceivedDelta(sal, FlagCritical); err != nil {
		t.Fatalf("expected no error, got %v", err)
	}
}

// TestValidateReceivedDeltaAdditiveWithoutCriticalAccepts verifies a
// non-REPLACE delta does not require FlagCritical.
func TestValidateReceivedDeltaAdditiveWithoutCriticalAccepts(t *testing.T) {
	sal := "A:ASD:DELTA[15.0→15.1:H+[LACTATE]]"
	if err := ValidateReceivedDelta(sal, 0); err != nil {
		t.Fatalf("expected no error for additive delta, got %v", err)
	}
}

// TestValidateReceivedDeltaNonDeltaPassthrough verifies non-DELTA SAL is a
// pass-through (validator is scoped to A:ASD:DELTA traffic).
func TestValidateReceivedDeltaNonDeltaPassthrough(t *testing.T) {
	for _, sal := range []string{"H:HR@NODE1", "A:ASD[15.1]", "A:ASD?"} {
		if err := ValidateReceivedDelta(sal, 0); err != nil {
			t.Fatalf("non-DELTA passthrough failed for %q: %v", sal, err)
		}
	}
}
